src/bot.rs

```diff
@@ -0,0 +1,175 @@
+        } else {
+            client.send_message(("Gioco non trovato!", from.id).into());
+        }
+    }
+}
+
+fn init_game(
+    game_index: usize,
+    from: telegram_bot_raw::types::chat::User,
+    playable_games: &Vec<Box<dyn Game>>,
+    player_games: &mut HashMap<telegram_bot_raw::types::refs::UserId, String>,
+    game_channel: &mut HashMap<String, std::sync::mpsc::SyncSender<threading::ThreadMessage>>,
+    game_last_played: &mut HashMap<String, std::time::Instant>,
+    client: &mut Telegram
+) {
+    use threading::ThreadMessage;
+    let game_id = ulid::Ulid::new().to_string();
+    let (sender, reciever) = mpsc::sync_channel(10);
+    sender.send(ThreadMessage::AddPlayer(primitives::Player{id: from.id.into(), name: utils::get_user_name(&from.first_name, &from.last_name)})).unwrap();
+    player_games.insert(from.id, game_id.clone());
+    game_channel.insert(game_id.clone(), sender);
+    game_last_played.insert(game_id.clone(), std::time::Instant::now());
+    client.send_message((format!("Per invitare altre persone condividi questo link: https://t.me/giocoacartebot?start={}", game_id), from.id).into());
+    let game_tg_client = client.clone();
+    game_agent::new_agent(game_tg_client, game_index, playable_games, reciever);
+}
+
+fn handle_callback_query(
+    qry: telegram_bot_raw::types::callback_query::CallbackQuery,
+    playable_games: &Vec<Box<dyn Game>>,
+    player_games: &mut HashMap<telegram_bot_raw::types::refs::UserId, String>,
+    game_channel: &mut HashMap<String, std::sync::mpsc::SyncSender<threading::ThreadMessage>>,
+    game_last_played: &mut HashMap<String, std::time::Instant>,
+    client: &mut Telegram
+) {
+    use threading::*;
+    //let qry_id: String = qry.id.into;
+    //client.ack_callback_query(&format!("{}", qry.id));
+    let data: Vec<String> = qry.data.unwrap().split(":").map(|x| x.to_owned()).collect();
+    let command = data[0].as_str();
+    match command {
+        "init_game" => {
+            let params = data[1].clone();
+            let index = params.parse::<usize>().unwrap();
+            init_game(index, qry.from, playable_games, player_games, game_channel, game_last_played, client);
+        },
+        "start" => {
+            let player_id = qry.from.id;
+            if let Some(game_id) = player_games.get(&player_id) {
+                if let Some(inst) = game_last_played.get_mut(game_id) {
+                    *inst = std::time::Instant::now();
+                }
+                let channel = game_channel.get(game_id).unwrap();
+                channel.send(ThreadMessage::Start).expect("Could not start game");
+            } else {
+                client.send_message(("Gioco non trovato", player_id).into());
+            }
+        },
+        "handle_move" => {
+            let card: primitives::Card = bincode::deserialize(&base64::decode(&data[1]).unwrap()).unwrap();
```

Comment:
I would make a function for the entire case "handle_move"

Reply:
Even if there's not much going on?

Reply:
yep